src/SecTester.Scan/Scan.cs

```diff
@@ -1,58 +1,160 @@
+public class Scan : IAsyncDisposable
 {
+  private static readonly TimeSpan DefaultPollingInterval = TimeSpan.FromSeconds(5);
+
   private static readonly IReadOnlyCollection<ScanStatus> ActiveStatuses =
-    new[]
-    {
-      ScanStatus.Pending, ScanStatus.Running, ScanStatus.Queued
-    };
+    new[] { ScanStatus.Pending, ScanStatus.Running, ScanStatus.Queued };
 
   private static readonly IReadOnlyCollection<ScanStatus> DoneStatuses =
-    new[]
-    {
-      ScanStatus.Disrupted, ScanStatus.Done, ScanStatus.Failed, ScanStatus.Stopped
-    };
+    new[] { ScanStatus.Disrupted, ScanStatus.Done, ScanStatus.Failed, ScanStatus.Stopped };
 
   private readonly ScanOptions _options;
-
+  private readonly ILogger _logger;
   private readonly Scans _scans;
-  private readonly ScanState _scanState = new(ScanStatus.Pending);
+  private ScanState _scanState = new(ScanStatus.Pending);
+  private readonly SemaphoreSlim _semaphore = new(1, 1);
+
+  public string Id { get; }
+
+  private bool DoneCore => DoneStatuses.Contains(_scanState.Status);
+  private bool ActiveCore => ActiveStatuses.Contains(_scanState.Status);
 
-  public Scan(Scans scans, ScanOptions options)
+  public bool Active
   {
+    get
+    {
+      using var _ = _semaphore.Lock();
+      return ActiveCore;
+    }
+  }
+
+  public bool Done
+  {
+    get
+    {
+      using var _ = _semaphore.Lock();
+      return DoneCore;
+    }
+  }
+
+  public Scan(string id, Scans scans, ILogger logger, ScanOptions options)
+  {
+    Id = id ?? throw new ArgumentNullException(nameof(id));
     _scans = scans ?? throw new ArgumentNullException(nameof(scans));
     _options = options ?? throw new ArgumentNullException(nameof(options));
+    _logger = logger ?? throw new ArgumentNullException(nameof(logger));
   }
 
-  public bool Active => ActiveStatuses.Contains(_scanState.Status);
+  public async ValueTask DisposeAsync()
+  {
+    try
+    {
+      await DisposeAsyncCore().ConfigureAwait(false);
+    }
+    catch
+    {
+      // ignore
+    }
 
-  public bool Done => DoneStatuses.Contains(_scanState.Status);
+    _semaphore.Dispose();
 
-  public void Dispose()
-  {
     GC.SuppressFinalize(this);
   }
 
-  public Task<IEnumerable<Issue>> Issues(CancellationToken cancellationToken = default)
+  protected virtual async ValueTask DisposeAsyncCore()
   {
-    throw new NotImplementedException();
+    if (_options.DeleteOnDispose is true)
+    {
+      await _scans.DeleteScan(Id).ConfigureAwait(false);
+    }
   }
 
-  public Task Stop()
+  public Task<IEnumerable<Issue>> Issues()
   {
-    throw new NotImplementedException();
+    return _scans.ListIssues(Id);
   }
 
-  public IAsyncEnumerable<ScanState> Status(CancellationToken cancellationToken = default)
+  public async Task Stop()
   {
-    throw new NotImplementedException();
+    try
+    {
+      using var _ = await _semaphore.LockAsync().ConfigureAwait(false);
+
+      await RefreshStateCore().ConfigureAwait(false);
```

Comment:
You can easily replace this with `RefreshState`. It looks like `RefreshStateCore` is useless